### PR TITLE
Adds Micrometer Observation instrumentation

### PR DESCRIPTION
**Pull request by marcingrzejszczak**
### Add an agent plugin to support <framework name>
- [X] Add a test case for the new plugin, refer to [the doc](https://github.com/apache/skywalking-java/blob/main/docs/en/setup/service-agent/java-agent/Plugin-test.md)
- [X] Add a component id in [the component-libraries.yml](https://github.com/apache/skywalking/blob/master/oap-server/server-starter/src/main/resources/component-libraries.yml) - https://github.com/apache/skywalking/pull/10078
- [X] Add a logo in [the UI repo](https://github.com/apache/skywalking-rocketbot-ui/tree/master/src/views/components/topology/assets) - https://github.com/apache/skywalking-booster-ui/pull/201

- [ ] If this pull request closes/resolves/fixes an existing issue, replace the issue number. Closes #<issue number>.
- [x] Update the [`CHANGES` log](https://github.com/apache/skywalking-java/blob/main/CHANGES.md).

### CODE REVIEW

**Comment by wu-sheng** [comment]
@pg-yang @yswdqz Does any of you have time to continue on this one? Marcin finished most parts.

**Comment by yswdqz** [comment]
> @pg-yang @yswdqz Does any of you have time to continue on this one? Marcin finished most parts.

Sorry, I don't have time recently.

**Comment by marcingrzejszczak** [comment]
The only thing to look at is the e2e plugin tests and modify the `expectedData.yml` for the tests to pass.

Also I see that we're currently in Spring naming spans in a different way than skywalking does. Should we leave it like this or rename the spans in the way that Skywalking does it?

**Comment by wu-sheng** [comment]
> Also I see that we're currently in Spring naming spans in a different way than skywalking does. Should we leave it like this or rename the spans in the way that Skywalking does it?

It is your call. URI style or `.`/spacing are helpful for endpoint searching. The endpoint is from span name of entry span.

**Comment by marcingrzejszczak** [comment]
Ok, the user can rename the observations at runtime by providing their own `ObservationFilter` to the registry, so that is not a problem. 

So the only thing that remains is to make the e2e pass and we're done.

File: test/plugin/scenarios/resttemplate-6.x-scenario/config/expectedData.yaml [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    - operationName: http get
+      parentSpanId: 0
+      spanId: 1
+      spanLayer: Http
+      startTime: nq 0
+      endTime: nq 0
+      componentId:
+      isError: false
+      spanType: Exit
+      peer: http://localhost:8080/resttemplate-6.x-scenario/resttemplate/syncback
```
Comment (wu-sheng) [inline test/plugin/scenarios/resttemplate-6.x-scenario/config/expectedData.yaml:65]: Peer should be remote addr only. Using `http.url` is working, but it increases the backend analysis payload a lot.

File: test/plugin/scenarios/resttemplate-6.x-scenario/config/expectedData.yaml [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+      spanType: Entry
+      peer: '/resttemplate-6.x-scenario/resttemplate/case/healthcheck'
+      tags:
+      - {key: http.url, value: '/resttemplate-6.x-scenario/resttemplate/case/healthcheck'}
+      - {key: method, value: HEAD}
+      - {key: status, value: '200'}
+      skipAnalysis: 'false'
+  - segmentId: not null
+    spans:
+    - operationName: http get /resttemplate/syncback
```
Comment (wu-sheng) [inline test/plugin/scenarios/resttemplate-6.x-scenario/config/expectedData.yaml:39]: We should check the ref in this span.

File: test/plugin/scenarios/resttemplate-6.x-scenario/config/expectedData.yaml [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+segmentItems:
+- serviceName: resttemplate-6.x-scenario
+  segmentSize: ge 2
+  segments:
+  - segmentId: not null
+    spans:
+    - operationName: http head /resttemplate/case/healthcheck
```
Comment (wu-sheng) [inline test/plugin/scenarios/resttemplate-6.x-scenario/config/expectedData.yaml:22]: `/healthcheck` is not required to verify.

**Comment by marcingrzejszczak** [comment]
I've managed to finally set up the e2e scenario with Spring Framework 6, jdk 17. 

In Skywalking got the following graph:

image

The headers get propagated

```
Got following headers [sw8:"1-ZDA3YTFjNzc4YWEzNDc3NWFmY2Q2MjQ3ZWY0N2ZiMzIuNTYuMTY2OTk5MzEzNzk1NjAwMDE=-ZDA3YTFjNzc4YWEzNDc3NWFmY2Q2MjQ3ZWY0N2ZiMzIuNTYuMTY2OTk5MzEzNzk1NjAwMDA=-1-WW91cl9BcHBsaWNhdGlvbk5hbWU=-MmFiMWNkNWRmZTJmNDJhNDkzNTA3ZjFkNDhiZTJmMTFAMTAuNy4xOC4xOTE=-aHR0cC5zZXJ2ZXIucmVxdWVzdHM=-bG9jYWxob3N0OjIzNDU=", sw8-correlation:"", sw8-x:"0-", accept:"text/plain, application/json, application/*+json, */*", user-agent:"Java/17.0.1", host:"localhost:2345", connection:"keep-alive"]
```

And the metrics are collected

```
 - Metric type  [TIMER],        name [http.client.requests],    tags [tag(error=none), tag(exception=none), tag(method=GET), tag(outcome=SUCCESS), tag(status=200), tag(uri=http://localhost:2345/resttemplate-6.x-scenario/resttemplate/syncback)],      measurements [Measurement{statistic='COUNT', value=1.0}, Measurement{statistic='TOTAL_TIME', value=123511.0}, Measurement{statistic='MAX', value=123511.0}]
 - Metric type  [LONG_TASK_TIMER],      name [http.server.requests.active],     tags [tag(exception=none), tag(method=GET), tag(outcome=SUCCESS), tag(status=200), tag(uri=UNKNOWN)],     measurements [Measurement{statistic='ACTIVE_TASKS', value=0.0}, Measurement{statistic='DURATION', value=0.0}]
 - Metric type  [TIMER],        name [http.server.requests],    tags [tag(error=none), tag(exception=none), tag(method=GET), tag(outcome=SUCCESS), tag(status=200), tag(uri=/resttemplate/syncback)],     measurements [Measurement{statistic='COUNT', value=1.0}, Measurement{statistic='TOTAL_TIME', value=68701.0}, Measurement{statistic='MAX', value=68701.0}]
 - Metric type  [LONG_TASK_TIMER],      name [http.client.requests.active],     tags [tag(exception=none), tag(method=GET), tag(outcome=UNKNOWN), tag(status=CLIENT_ERROR), tag(uri=http://localhost:2345/resttemplate-6.x-scenario/resttemplate/syncback)],      measurements [Measurement{statistic='ACTIVE_TASKS', value=0.0}, Measurement{statistic='DURATION', value=0.0}]
 - Metric type  [TIMER],        name [http.server.requests],    tags [tag(error=none), tag(exception=none), tag(method=GET), tag(outcome=SUCCESS), tag(status=200), tag(uri=/resttemplate/case/resttemplate)],    measurements [Measurement{statistic='COUNT', value=1.0}, Measurement{statistic='TOTAL_TIME', value=149568.0}, Measurement{statistic='MAX', value=149568.0}]
```

**Comment by marcingrzejszczak** [comment]
Skywalking has nothing to do with Sleuth. There is no Sleuth meter config, there's only micrometer meter config.

File: apm-sniffer/apm-toolkit-activation/apm-toolkit-micrometer-activation/src/main/java/org/apache/skywalking/apm/toolkit/activation/micrometer/MicrometerDefaultTracingHandlerInstrumentation.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+    public static final String INTERCEPT_START_POINT_METHOD = "onStart";
+
+    public static final String INTERCEPT_STOP_POINT_METHOD = "onStop";
+
+    public static final String INTERCEPT_ERROR_POINT_METHOD = "onError";
+
+    public static final String INTERCEPT_EVENT_POINT_METHOD = "onEvent";
+
+    public static final String INTERCEPT_CLASS = "org.apache.skywalking.apm.plugin.micrometer.MicrometerDefaultTracingHandlerInterceptor";
```
Comment (mrproliu) [inline apm-sniffer/apm-toolkit-activation/apm-toolkit-micrometer-activation/src/main/java/org/apache/skywalking/apm/toolkit/activation/micrometer/MicrometerDefaultTracingHandlerInstrumentation.java:43]: The interceptor classes of all instrumentation are not correct, please help to fix it. 
In this case, it should be: `org.apache.skywalking.apm.toolkit.activation.micrometer.MicrometerDefaultTracingHandlerInterceptor`

File: apm-application-toolkit/apm-toolkit-micrometer-registry/src/main/java/org/apache/skywalking/apm/toolkit/micrometer/observation/SkywalkingContextSnapshotThreadLocalAccessor.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+ */
+
+package org.apache.skywalking.apm.toolkit.micrometer.observation;
+
+import io.micrometer.context.ThreadLocalAccessor;
+
+/**
+ * A {@link ThreadLocalAccessor} to put and restore current {@code ContextSnapshot} from APM agent.
+ */
+public class SkywalkingContextSnapshotThreadLocalAccessor implements ThreadLocalAccessor<Object> {
```
[the remark this answers is not in the document]
Comment (marcingrzejszczak) [reply apm-application-toolkit/apm-toolkit-micrometer-registry/src/main/java/org/apache/skywalking/apm/toolkit/micrometer/observation/SkywalkingContextSnapshotThreadLocalAccessor.java:26]: Actually I would need some feedback from you. When we continue a span between threads, what should be the name of that continued span? Also how can you reset the thread local with a span (how can I remove the active span from thread local)

File: apm-sniffer/apm-toolkit-activation/apm-toolkit-micrometer-activation/src/main/java/org/apache/skywalking/apm/toolkit/activation/micrometer/MicrometerDefaultTracingHandlerInstrumentation.java (hunk 4 above)
Comment (marcingrzejszczak) [reply apm-sniffer/apm-toolkit-activation/apm-toolkit-micrometer-activation/src/main/java/org/apache/skywalking/apm/toolkit/activation/micrometer/MicrometerDefaultTracingHandlerInstrumentation.java:43]: I guess I'll need to move the classes around again after talking to @wu-sheng but thanks for pointing this out

File: apm-application-toolkit/apm-toolkit-micrometer-registry/src/main/java/org/apache/skywalking/apm/toolkit/micrometer/observation/SkywalkingContextSnapshotThreadLocalAccessor.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (wu-sheng) [reply apm-application-toolkit/apm-toolkit-micrometer-registry/src/main/java/org/apache/skywalking/apm/toolkit/micrometer/observation/SkywalkingContextSnapshotThreadLocalAccessor.java:26]: Usually, we don't continue a span in a new thread. We only support finishing a span in a new thread, or continuous tracing context(capture and continuous snapshot from ContextManager)

[the remark this answers is not in the document]
Comment (mrproliu) [reply apm-application-toolkit/apm-toolkit-micrometer-registry/src/main/java/org/apache/skywalking/apm/toolkit/micrometer/observation/SkywalkingContextSnapshotThreadLocalAccessor.java:26]: > When we continue a span between threads, what should be the name of that continued span? 

It's usually decided by when you need to cross-thread. Such as when we submit a new thread, we called `Thread/{thread_name}/{running_function}`.

> Also how can you reset the thread local with a span (how can I remove the active span from thread local)

Here has a good [demo](https://github.com/apache/skywalking-java/blob/main/apm-sniffer/apm-toolkit-activation/apm-toolkit-trace-activation/src/main/java/org/apache/skywalking/apm/toolkit/activation/trace/CallableOrRunnableActivation.java), you could check this out. In this demo, we capture a snapshot from the tracing context and continue the snapshot in other thread.

**Review by mrproliu — APPROVED** [review]

**Review by wu-sheng — APPROVED** [review]

**Comment by wu-sheng** [comment]
@marcingrzejszczak We have 8.14.0 released now. Could you try to make your plugin test back to the CI pipeline?

**Comment by marcingrzejszczak** [comment]
I will do it asap, thanks :)

**Pull request merged**